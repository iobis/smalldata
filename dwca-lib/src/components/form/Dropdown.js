import classNames from 'classnames'
import PropTypes from 'prop-types'
import React, { useEffect, useRef, useState } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { useOnClickOutside } from '@smalldata/dwca-lib'

export default function Dropdown({ onChange, options, value }) {
  const ref = useRef()
  const [active, setActive] = useState(false)
  const [selectedValue, setSelectedValue] = useState(value)
  const hideOptions = () => setActive(false)
  const showOptions = () => {if (active === false) setActive(true)}
  const handleChange = (item) => {
    onChange(item)
    setSelectedValue(item)
    setActive(false)
  }

  useOnClickOutside(ref, hideOptions)

  useEffect(() => {
    setSelectedValue(value)
  }, [value])

  return (
    <div className={classNames('dropdown', { 'is-active': active })} onClick={showOptions} ref={ref}>
      <div className="dropdown-trigger">
        <button aria-controls="dropdown-menu" aria-haspopup="true" className="button">
          <span className="selected-value">{selectedValue}</span>
          <span className="icon is-small">
            <FontAwesomeIcon className="angle-down" icon="angle-down"/>
          </span>
        </button>
      </div>
      <div className="dropdown-menu" id="dropdown-menu" role="menu">
        <div className="dropdown-content">
          {options.map(option => (
            <div
              className={classNames('dropdown-item', { 'is-active': option === selectedValue })}
              key={option}
              onClick={() => handleChange(option)}>
              {option}
            </div>
          ))}
        </div>
      </div>
    </div>
  )
}

Dropdown.propTypes = {
  onChange: PropTypes.func.isRequired,
  options:  PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  value:    PropTypes.string.isRequired
}
