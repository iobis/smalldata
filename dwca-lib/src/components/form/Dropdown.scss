.dropdown {
  display: block;
  .dropdown-item {
    cursor: pointer;

    &:hover {
      background-color: $dropdown-item-hover-background-color;
    }

    &.is-active {
      background-color: $dropdown-item-active-background-color;
      color: $dropdown-item-active-color;
    }
  }
  .dropdown-trigger {
    width: 80%;
  }
}
