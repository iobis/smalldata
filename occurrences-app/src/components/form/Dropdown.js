import classNames from 'classnames'
import PropTypes from 'prop-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { useOnClickOutside } from '../../hooks/hooks'
import React, { useRef, useState } from 'react'

export default function Dropdown({ onChange, options, value }) {
  const ref = useRef()
  const [active, setActive] = useState(false)
  const [selectedValue, setSelectedValue] = useState(value)
  const hideOptions = () => setActive(false)
  const showOptions = () => {if (active === false) setActive(true)}
  const handleChange = (item) => {
    onChange(item)
    setSelectedValue(item)
    setActive(false)
  }

  useOnClickOutside(ref, hideOptions)

  return (
    <div className={classNames('dropdown', { 'is-active': active })} onClick={showOptions} ref={ref}>
      <div className="dropdown-trigger">
        <button aria-controls="dropdown-menu" aria-haspopup="true" className="button">
          <span className="selected-value">{selectedValue}</span>
          <span className="icon is-small">
            <FontAwesomeIcon className="angle-down" icon="angle-down"/>
          </span>
        </button>
      </div>
      <div className="dropdown-menu" id="dropdown-menu" role="menu">
        <div className="dropdown-content">
          {options.map(option => (
            <a
              className={classNames('dropdown-item', { 'is-active': option === selectedValue })}
              href="#"
              key={option}
              onClick={() => handleChange(option)}>
              {option}
            </a>
          ))}
        </div>
      </div>
    </div>
  )
}

Dropdown.propTypes = {
  onChange: PropTypes.func.isRequired,
  options:  PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  value:    PropTypes.string.isRequired
}
