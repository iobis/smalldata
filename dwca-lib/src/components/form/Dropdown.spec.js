import Dropdown from './Dropdown'
import React from 'react'
import { mount } from 'enzyme'

describe('Dropdown', () => {
  it('renders correctly', () => {
    expect(mount(createComponent())).toMatchSnapshot()
  })

  it('returns selected value when changing', () => {
    const onChange = jest.fn()
    const wrapper = mount(createComponent({ onChange }))

    wrapper.find('.dropdown-trigger button').simulate('click')
    expect(wrapper.find('.selected-value').text()).toBe('option-1')
    expect(wrapper.find('.dropdown').hasClass('is-active')).toBe(true)
    expect(wrapper.find('.dropdown-item')).toHaveLength(3)
    expect(onChange).toHaveBeenCalledTimes(0)

    wrapper.find('.dropdown-item').at(1).simulate('click')
    expect(wrapper.find('.selected-value').text()).toBe('option-2')
    expect(wrapper.find('.dropdown').hasClass('is-active')).toBe(false)
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toBeCalledWith('option-2')
  })

  it('updates selected value when props changing', () => {
    const onChange = jest.fn()
    const wrapper = mount(createComponent({ onChange }))

    wrapper.find('.dropdown-trigger button').simulate('click')
    expect(wrapper.find('.selected-value').text()).toBe('option-1')
    expect(wrapper.find('.dropdown').hasClass('is-active')).toBe(true)
    expect(wrapper.find('.dropdown-item')).toHaveLength(3)
    expect(onChange).toHaveBeenCalledTimes(0)

    wrapper.setProps({ value: 'option-3' })
    expect(wrapper.find('.selected-value').text()).toBe('option-3')
  })
})

function createComponent(props) {
  const defaultProps = {
    options:  ['option-1', 'option-2', 'option-3'],
    onChange: jest.fn(),
    value:    'option-1'
  }
  return <Dropdown {...defaultProps} {...props}/>
}
