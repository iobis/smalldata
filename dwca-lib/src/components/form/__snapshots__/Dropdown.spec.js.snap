// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly 1`] = `
<Dropdown
  onChange={[MockFunction]}
  options={
    Array [
      "option-1",
      "option-2",
      "option-3",
    ]
  }
  value="option-1"
>
  <div
    className="dropdown"
  >
    <div
      className="dropdown-trigger"
    >
      <button
        aria-controls="dropdown-menu"
        aria-haspopup="true"
        className="button"
        onClick={[Function]}
      >
        <span
          className="selected-value"
        >
          option-1
        </span>
        <span
          className="icon is-small"
        >
          <FontAwesomeIcon
            className="angle-down"
            icon="angle-down"
          >
            <i
              className="angle-down"
              icon="angle-down"
            />
          </FontAwesomeIcon>
        </span>
      </button>
    </div>
    <div
      className="dropdown-menu"
      id="dropdown-menu"
      role="menu"
    >
      <div
        className="dropdown-content"
      >
        <div
          className="dropdown-item is-active"
          key="option-1"
          onClick={[Function]}
        >
          option-1
        </div>
        <div
          className="dropdown-item"
          key="option-2"
          onClick={[Function]}
        >
          option-2
        </div>
        <div
          className="dropdown-item"
          key="option-3"
          onClick={[Function]}
        >
          option-3
        </div>
      </div>
    </div>
  </div>
</Dropdown>
`;
